src/opentimelineio/composition.h

```diff
@@ -72,12 +78,170 @@ class Composition : public Item {
     // XXX: python implementation is O(n^2) in number of children
     std::vector<Composable*> _children_at_time(RationalTime, ErrorStatus* error_status) const;
 
+    size_t _bisect_right(
+        RationalTime const& tgt,
+        std::function<RationalTime(Composable*)> const& key_func,
+        ErrorStatus* error_status,
+        optional<size_t> lower_search_bound = optional<size_t>(0),
+        optional<size_t> upper_search_bound = nullopt) const;
+    size_t _bisect_left(
+        RationalTime const& tgt,
+        std::function<RationalTime(Composable*)> const& key_func,
+        ErrorStatus* error_status,
+        optional<size_t> lower_search_bound = optional<size_t>(0),
+        optional<size_t> upper_search_bound = nullopt) const;
+
     std::vector<Retainer<Composable>> _children;
     
     // This is for fast lookup only, and varies automatically
     // as _children is mutated.
     std::set<Composable*> _child_set;
 };
 
+template<typename T>
+inline std::vector<SerializableObject::Retainer<T>> Composition::each_child(
+    ErrorStatus* error_status,
+    optional<TimeRange> search_range,
+    bool shallow_search) const
+{
+    std::vector<Retainer<Composable>> children;
+    if (search_range)
+    {
+        auto range_map = range_of_all_children(error_status);
```

Comment:
I'd recommend busting this out to a separate non-templated "filter" function. It's a generally useful function, likely to be used more than once. Busting out would reduce the amount of per-class code instantiated per <T>.

Reply:
In composition.py there is also a function called child_at_time() which is sort of similar, but for a specific time instead of a time range.

Reply:
child_at_time() and children_in_range() perhaps?